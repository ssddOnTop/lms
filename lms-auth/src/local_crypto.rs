use totp_rs::TOTP;
use anyhow::Result;
use base64::Engine;
use base64::prelude::BASE64_STANDARD;
use sha2::{Digest, Sha256};

const IV: [u8; 16] = [7; 16];

pub fn gen_totp(totp: &TOTP) -> Result<String> {
    Ok(totp.generate_current()?)
}

pub fn verify_totp(totp: &TOTP, code: &str) -> Result<bool> {
    Ok(totp.check_current(code)?)
}

pub fn encrypt_aes(key: &[u8], data: &str) -> Result<String> {
    let key: &[u8; 16] = key.try_into()?;
    let cipher  = libaes::Cipher::new_128(key);
    let encrypted  = cipher.cbc_encrypt(&IV, data.as_bytes());
    let encrypted = BASE64_STANDARD.encode(encrypted);
    Ok(encrypted)
}

pub fn decrypt_aes<T: AsRef<[u8]>>(key: &[u8], data: T) -> Result<String> {
    let key: &[u8; 16] = key.try_into()?;
    let cipher  = libaes::Cipher::new_128(key);
    let data = BASE64_STANDARD.decode(data.as_ref())?;
    let decrypted = cipher.cbc_decrypt(&IV, &data);
    Ok(String::from_utf8(decrypted)?)
}

pub fn hash_256<T: AsRef<str>>(data: T) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data.as_ref().as_bytes());
    let data = hasher.finalize();
    let data = hex::encode(data);
    data
}

#[cfg(test)]
mod tests {
    use totp_rs::{Algorithm, Secret};
    use super::*;

    #[test]
    fn test_encrypt_decrypt() {
        let key = [1; 16];
        let data = "hello world";
        let encrypted = encrypt_aes(&key, data).unwrap();
        let decrypted = decrypt_aes(&key, &encrypted).unwrap();
        assert_eq!(data, decrypted);
    }

    #[test]
    fn test_gen_verify_totp() {
        let totp = TOTP::new(
            Algorithm::SHA1,
            6,
            1,
            30,
            Secret::Raw("JBSWY3DPEHPK3PXP".as_bytes().to_vec()).to_bytes().unwrap(),
        ).unwrap();

        let generated_totp = gen_totp(&totp).unwrap();

        let is_valid = verify_totp(&totp, &generated_totp).unwrap();

        assert!(is_valid);
    }

    #[test]
    fn test_hash_256() {
        let data = "hello world";
        let hashed = hash_256(data);
        assert_eq!(hashed, "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9");
    }
}