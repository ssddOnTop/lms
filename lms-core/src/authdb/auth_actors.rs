use anyhow::anyhow;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::str::FromStr;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub enum Authority {
    Admin,
    Faculty,
    Student,
}

impl FromStr for Authority {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_ascii_lowercase();
        match s.as_str() {
            "admin" => Ok(Authority::Admin),
            "faculty" => Ok(Authority::Faculty),
            "student" => Ok(Authority::Student),
            _ => Err(anyhow!("Unable to serialize")),
        }
    }
}

impl Authority {
    pub fn from_int(int: u8) -> anyhow::Result<Self> {
        match int {
            0 => Ok(Authority::Admin),
            1 => Ok(Authority::Faculty),
            2 => Ok(Authority::Student),
            _ => Err(anyhow!("Unable to serialize")),
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct User {
    pub username: String,
    pub name: String,
    pub password: String,
    pub authority: Authority,
}

#[derive(Clone, Default, Debug, Serialize, Deserialize, PartialEq)]
pub struct Users {
    users: HashMap<String, User>,
}

impl Users {
    pub fn get(&self, username: &str) -> Option<User> {
        self.users.get(username).cloned()
    }
    pub fn insert(&mut self, user: User) {
        self.users.insert(user.username.clone(), user);
    }
}

#[cfg(test)]
mod tests {
    use crate::authdb::auth_actors::{Authority, User, Users};
    use std::collections::HashMap;

    #[test]
    fn test_ser() -> anyhow::Result<()> {
        let mut users = HashMap::new();
        users.insert(
            "foo".to_string(),
            User {
                username: "foo".to_string(),
                name: "Foo".to_string(),
                password: "foopassword".to_string(),
                authority: Authority::Admin,
            },
        );
        let users = Users { users };

        let ser = serde_json::to_string(&users)?;
        assert_eq!(
            ser,
            r#"{"users":{"foo":{"username":"foo","name":"Foo","password":"foopassword","authority":"Admin"}}}"#
        );
        Ok(())
    }
    #[test]
    fn test_deser() -> anyhow::Result<()> {
        let ser = r#"{"users":{"foo":{"username":"foo","name":"Foo","password":"foopassword","authority":"Admin"}}}"#;
        let mut users = HashMap::new();
        users.insert(
            "foo".to_string(),
            User {
                username: "foo".to_string(),
                name: "Foo".to_string(),
                password: "foopassword".to_string(),
                authority: Authority::Admin,
            },
        );
        let users = Users { users };
        assert_eq!(users, serde_json::from_str(ser)?);
        Ok(())
    }
}
