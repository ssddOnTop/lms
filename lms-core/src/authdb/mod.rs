pub mod auth_actors;
pub mod auth_db;
