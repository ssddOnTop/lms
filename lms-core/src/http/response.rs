use anyhow::{Context, Result};
use bytes::Bytes;
use derive_setters::Setters;
use http_body_util::Full;
use serde::de;
use std::str::FromStr;

#[derive(Clone, Debug, Default, Setters)]
pub struct Response<Body: Default + Clone> {
    pub status: reqwest::StatusCode,
    pub headers: reqwest::header::HeaderMap,
    pub body: Body,
}

impl Response<Bytes> {
    pub async fn from_reqwest(resp: reqwest::Response) -> Result<Self> {
        let status = resp.status();
        let headers = resp.headers().to_owned();
        let body = resp.bytes().await?;
        Ok(Response {
            status,
            headers,
            body,
        })
    }
    pub fn empty() -> Self {
        Response {
            status: reqwest::StatusCode::OK,
            headers: reqwest::header::HeaderMap::default(),
            body: Bytes::new(),
        }
    }

    pub fn to_json<T: de::DeserializeOwned + Clone + Default>(self) -> Result<Response<T>> {
        let mut resp = Response::default();
        let body = serde_json::from_slice::<T>(&self.body)?;
        resp.body = body;
        resp.status = self.status;
        resp.headers = self.headers;
        Ok(resp)
    }

    pub fn to_resp_string(self) -> Result<Response<String>> {
        Ok(Response::<String> {
            body: String::from_utf8(self.body.to_vec())?,
            status: self.status,
            headers: self.headers,
        })
    }
    pub fn into_hyper(self) -> Result<hyper::Response<Full<Bytes>>> {
        let mut builder =
            hyper::Response::builder().status(hyper::StatusCode::from_u16(self.status.as_u16())?);
        for (key, value) in self.headers {
            builder = builder.header(
                hyper::header::HeaderName::from_str(
                    key.context("Invalid header key")?
                        .as_str()
                        .to_string()
                        .as_str(),
                )?,
                hyper::header::HeaderValue::from_str(value.to_str()?)?,
            );
        }
        Ok(builder.body(Full::new(self.body))?)
    }
}
