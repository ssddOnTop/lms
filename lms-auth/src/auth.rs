use crate::is_default;
use anyhow::{anyhow, Context, Result};
use http_body_util::Full;
use libaes::AES_256_KEY_LEN;
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use totp_rs::TOTP;

use crate::local_crypto::{decrypt_aes, encrypt_aes, gen_totp, hash_128, hash_256};

#[derive(Debug, Clone)]
pub struct AuthProvider {
    auth_db_path: String,
    totp: TOTP,
    aes_key: Vec<u8>,
}

pub enum RequestType {
    Login,
    Signup,
}

impl Display for RequestType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            RequestType::Login => f.write_str("login"),
            RequestType::Signup => f.write_str("signup"),
        }
    }
}

#[derive(Serialize, Deserialize)]
pub struct AuthRequest {
    pub username: String,
    pub password: String,
    pub signature: String,
    #[serde(default, skip_serializing_if = "is_default")]
    pub signup_details: Option<SignUpDet>,
}

#[derive(PartialEq, Eq, Serialize, Deserialize)]
pub struct SignUpDet {
    pub name: String,
    pub authority: u8,
    pub admin_username: String,
    pub admin_password: String,
}

#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AuthResult {
    #[serde(default, skip_serializing_if = "is_default")]
    pub error: Option<AuthError>,

    #[serde(default, skip_serializing_if = "is_default")]
    pub success: Option<AuthSucc>,
}

#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AuthSucc {
    #[serde(default, skip_serializing_if = "is_default")]
    pub name: String,
    pub token: String,
}

#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AuthError {
    #[serde(default, skip_serializing_if = "is_default")]
    pub message: String,
}

impl AuthRequest {
    fn new<T: AsRef<str>>(
        username: T,
        password: T,
        provider: &AuthProvider,
        sign_up_det: Option<SignUpDet>,
    ) -> Result<Self> {
        let password = hash_256(password);
        let extra_hash = provider.gen_sig(username.as_ref(), &password)?;

        let mut signup_details = None;

        if let Some(mut signup_det) = sign_up_det {
            signup_det.admin_password = hash_256(signup_det.admin_password);
            signup_details = Some(signup_det);
        }

        Ok(Self {
            username: username.as_ref().to_string(),
            password,
            signature: extra_hash,
            signup_details,
        })
    }
    fn into_encrypted_request(self, auth_provider: &AuthProvider) -> Result<String> {
        let request = auth_provider
            .encrypt_aes(serde_json::to_string(&self)?)
            .map_err(|_| anyhow!("Unable to encrypt request"))?;
        Ok(request)
    }
    pub fn try_from_encrypted<T: AsRef<[u8]>>(
        req: T,
        auth_provider: &AuthProvider,
    ) -> Result<Self> {
        let req = auth_provider.decrypt_aes(req)?;
        println!("hx");
        let req = serde_json::from_str::<Self>(&req)?;
        Ok(req)
    }
    pub fn verify_sig(&self, auth_provider: &AuthProvider) -> bool {
        let sig = auth_provider.gen_sig(&self.username, &self.password);
        println!("L: {:?}", sig);
        println!("R: {}", self.signature);
        sig.map(|v| v.eq(&self.signature)).unwrap_or(false)
    }
}

impl AuthResult {
    pub fn try_from_encrypted_response(aes_key: &[u8], response: &str) -> Result<Self> {
        let response =
            decrypt_aes(aes_key, response).map_err(|_| anyhow!("Unable to decrypt response"))?;
        let result = serde_json::from_str::<AuthResult>(&response)?;
        Ok(result)
    }
    pub fn into_hyper_response(self) -> Result<hyper::Response<Full<bytes::Bytes>>> {
        let body = serde_json::to_string(&self)?;
        let response = hyper::Response::builder()
            .status(200)
            .header("Content-Type", "application/json")
            .body(Full::new(bytes::Bytes::from(body)))?;
        Ok(response)
    }
}

impl AuthProvider {
    pub fn init(auth_db_path: String, totp: TOTP, aes_key: String) -> Result<AuthProvider> {
        let aes_key = hash_128(aes_key)[..32].to_string();
        assert_eq!(aes_key.len(), AES_256_KEY_LEN, "AES key must be 16 bytes");

        let provider = Self {
            auth_db_path,
            totp,
            aes_key: aes_key.into_bytes(),
        };
        Ok(provider)
    }

    pub async fn authenticate(&self, username: &str, password: &str) -> Result<AuthSucc> {
        let request = AuthRequest::new(username, password, self, None)?;
        let request = request.into_encrypted_request(self)?;

        let response = reqwest::Client::new()
            .post(self.auth_db_path.clone())
            .body(request)
            .send()
            .await?;

        let result =
            AuthResult::try_from_encrypted_response(&self.aes_key, &response.text().await?)?;

        if let Some(err) = result.error {
            return Err(anyhow::anyhow!(err.message));
        }

        result.success.context("Internal error: Empty response")
    }
    pub fn encrypt_aes<T: AsRef<str>>(&self, content: T) -> Result<String> {
        encrypt_aes(&self.aes_key, content.as_ref())
    }

    pub fn decrypt_aes<T: AsRef<[u8]>>(&self, content: T) -> Result<String> {
        decrypt_aes(&self.aes_key, content.as_ref())
    }
    pub fn gen_sig(&self, a: &str, b: &str) -> Result<String> {
        let totp = gen_totp(&self.totp)?;
        Ok(hash_256(format!("{}ssdd{}{}", totp, a, b)))
    }
    pub fn db_path(&self) -> &str {
        &self.auth_db_path
    }
    pub fn get_pw(&self) -> &[u8] {
        &self.aes_key
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use totp_rs::{Algorithm, Secret};

    #[test]
    fn test_deser_auth_resp() {
        let json = r#"{
            "success": {
                "token": "token"
            }
        }"#;

        let json_error = r#"{
            "error": {
                "message": "error message"
            }
        }"#;

        let result = serde_json::from_str::<AuthResult>(json);
        let result_err = serde_json::from_str::<AuthResult>(json_error);

        assert!(result.unwrap().success.is_some());
        assert!(result_err.unwrap().error.is_some());
    }

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }

    #[tokio::test]
    async fn test_authenticate_success() -> Result<()> {
        let server = start_mock_server();

        let totp = TOTP::new(
            Algorithm::SHA1,
            6,
            1,
            30,
            Secret::Raw("JBSWY3DPEHPK3PXP".as_bytes().to_vec())
                .to_bytes()
                .unwrap(),
        )
        .unwrap();

        let aes_key = "12345678901234561234567890123456".to_string();
        let server_url = server.base_url();

        let provider = AuthProvider::init(server_url, totp, aes_key)?;

        let req = AuthRequest::new("user", "pass", &provider, None)?;
        let req = req.into_encrypted_request(&provider)?;
        let resp_json = json!({
            "success": {
                "name": "John Doe",
                "token": "abcdef123456"
            }
        });

        let resp = encrypt_aes(&provider.aes_key, &serde_json::to_string(&resp_json)?)?;

        let m = server.mock(|when, then| {
            when.method(httpmock::Method::POST).path("/").body(req);

            then.status(200).body(resp);
        });

        let result = provider.authenticate("user", "pass").await;

        m.assert();
        assert!(result.is_ok());

        let success = result.unwrap();
        assert_eq!(success.name, "John Doe");
        assert_eq!(success.token, "abcdef123456");

        Ok(())
    }

    /*    #[tokio::test]
    async fn test_req_gen() -> Result<()> { // TODO DROP THIS
        let totp = TOTP::new(
            Algorithm::SHA1,
            6,
            1,
            30,
            Secret::Raw("base32encodedkey".as_bytes().to_vec()).to_bytes()?,
        )?;
        let auth = AuthProvider::init(
            "/home/ssdd/RustroverProjects/lms/auth".to_string(),
            totp,
            hash_256("32bytebase64encodedkey"),
        )?;

        let signup = SignUpDet {
            name: "fooo".to_string(),
            authority: 0,
            admin_username: "admin".to_string(),
            admin_password: "tmppw".to_string(),
        };

        let req = AuthRequest::new("user", "pass", &auth, Some(signup))?;
        let resp = reqwest::Client::new()
            .post("http://localhost:19194/auth")
            .body(req.into_encrypted_request(&auth)?)
            .send()
            .await?;
        println!("{:?}", resp.text().await);
        Ok(())
    }*/
}
