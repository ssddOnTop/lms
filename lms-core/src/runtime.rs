use std::sync::Arc;

use crate::{FileIO, HttpIO, Instance};

/// The TargetRuntime struct unifies the available runtime-specific
/// IO implementations. This is used to reduce piping IO structs all
/// over the codebase.
#[derive(Clone)]
pub struct TargetRuntime {
    /// HTTP client for making standard HTTP requests.
    pub http: Arc<dyn HttpIO>,
    /// Interface for file operations, tailored to the target environment's
    /// capabilities.
    pub file: Arc<dyn FileIO>,

    /// Instance gives current time since epoch.
    pub instance: Arc<dyn Instance>,
}

#[cfg(test)]
pub mod tests {
    use std::sync::Arc;
    use std::time::SystemTime;

    use anyhow::{Context, Result};
    use dashmap::DashMap;
    use hyper::body::Bytes;
    use reqwest::Client;

    use crate::http::response::Response;
    use crate::runtime::TargetRuntime;
    use crate::{FileIO, HttpIO, Instance};

    #[derive(Default)]
    struct TestHttp {
        client: Client,
    }

    impl TestHttp {
        fn init() -> Arc<Self> {
            Arc::new(Self::default())
        }
    }

    #[async_trait::async_trait]
    impl HttpIO for TestHttp {
        async fn execute(&self, request: reqwest::Request) -> Result<Response<Bytes>> {
            let response = self.client.execute(request).await;
            Response::from_reqwest(
                response?
                    .error_for_status()
                    .map_err(|err| err.without_url())?,
            )
            .await
        }
    }

    #[derive(Clone)]
    struct TestFileIO {
        hm: DashMap<String, Vec<u8>>,
    }

    impl TestFileIO {
        fn init() -> Self {
            TestFileIO { hm: DashMap::new() }
        }
    }

    #[async_trait::async_trait]
    impl FileIO for TestFileIO {
        async fn write<'a>(&'a self, path: &'a str, content: &'a [u8]) -> anyhow::Result<()> {
            self.hm.insert(path.to_string(), content.to_vec());
            Ok(())
        }

        async fn read<'a>(&'a self, path: &'a str) -> anyhow::Result<String> {
            let buffer = self
                .hm
                .get(path)
                .context(format!("File: {} not found", path))?
                .clone();
            Ok(String::from_utf8(buffer)?)
        }

        async fn create_dirs<'a>(&'a self, _path: &'a str) -> Result<()> {
            Ok(())
        }
    }

    #[derive(Clone)]
    struct TestInstance {}

    impl Instance for TestInstance {
        fn now(&self) -> Result<u128> {
            Ok(SystemTime::now()
                .duration_since(SystemTime::UNIX_EPOCH)?
                .as_millis())
        }
    }

    pub fn init() -> TargetRuntime {
        let http = TestHttp::init();

        let file = TestFileIO::init();
        TargetRuntime {
            http,
            file: Arc::new(file),
            instance: Arc::new(TestInstance {}),
        }
    }
}
