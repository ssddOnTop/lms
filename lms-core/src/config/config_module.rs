use crate::authdb::auth_actors::Users;
use crate::config::Config;
use crate::runtime::TargetRuntime;
use hyper::Method;
use lms_auth::auth::AuthProvider;
use reqwest::{Body, Request};
use serde_json::json;
use std::ops::Deref;

#[derive(Default, Debug, Clone)]
pub struct ConfigModule {
    pub config: Config,
    pub extensions: Extensions,
}

#[derive(Default, Debug, Clone)]
pub struct Extensions {
    pub users: Option<Users>,
    pub auth: Option<AuthProvider>,
}

impl Deref for ConfigModule {
    type Target = Config;
    fn deref(&self) -> &Self::Target {
        &self.config
    }
}

impl From<Config> for ConfigModule {
    fn from(value: Config) -> Self {
        Self {
            config: value,
            extensions: Extensions {
                users: None,
                auth: None,
            },
        }
    }
}

impl ConfigModule {
    pub async fn resolve(self, target_runtime: &TargetRuntime) -> anyhow::Result<Self> {
        let totp = self.config.auth.totp.clone().into_totp()?;

        assert!(
            self.config.auth.aes_key.len() > 8,
            "db_file_password must be at least 8 characters long"
        );

        let auth = AuthProvider::init(
            self.config.auth.auth_db_path.clone(),
            totp,
            lms_auth::local_crypto::hash_256(&self.config.auth.aes_key).clone(),
        )?;

        let users = if !auth.db_path().starts_with("http") {
            match target_runtime.file.read(auth.db_path()).await {
                Ok(encrypted_users) => {
                    let decrypted_users = auth.decrypt_aes(encrypted_users)?;

                    serde_json::from_str::<Users>(&decrypted_users)?
                }
                Err(_) => {
                    let users = Users::default();
                    let encrypted_users = auth.encrypt_aes(serde_json::to_string(&users)?)?;
                    target_runtime
                        .file
                        .write(auth.db_path(), encrypted_users.as_bytes())
                        .await?;
                    users
                }
            }
        } else {
            let url = url::Url::parse(auth.db_path())?;

            let mut req = Request::new(Method::POST, url);
            let password = String::from_utf8(auth.get_pw().to_vec())?;
            *req.body_mut() = Some(Body::from(
                json!({
                    "operation": "get_users",
                    "pw": &password
                })
                .to_string(),
            ));
            let result = target_runtime.http.execute(req).await?;
            serde_json::from_slice::<Users>(&result.body)?
        };

        Ok(ConfigModule {
            extensions: Extensions {
                users: Some(users),
                auth: Some(auth),
            },
            ..self
        })
    }
}
