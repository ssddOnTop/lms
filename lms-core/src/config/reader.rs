#![allow(unused)]

use crate::config::config_module::ConfigModule;
use crate::config::Config;
use crate::runtime::TargetRuntime;
use reqwest::Url;
use std::path::Path;

/// Reads the configuration from a file or from an HTTP URL and resolves all linked extensions to create a ConfigModule.
pub struct ConfigReader {
    runtime: TargetRuntime,
}

/// Response of a file read operation
#[derive(Debug)]
struct FileRead {
    content: String,
    path: String,
}

impl ConfigReader {
    pub fn init(runtime: TargetRuntime) -> Self {
        Self { runtime }
    }

    /// Reads the config file and returns serialized config
    pub async fn read<T: AsRef<str>>(&self, file: T) -> anyhow::Result<ConfigModule> {
        let file = self.read_file(file).await?;
        let config = Config::from_json(&file.content)?;
        let parent = Path::new(&file.path).parent();
        let config_module = ConfigModule::from(config)
            .resolve(&self.runtime, parent)
            .await?;

        Ok(config_module)
    }
    /// Reads a file from the filesystem or from an HTTP URL
    async fn read_file<T: AsRef<str>>(&self, file: T) -> anyhow::Result<FileRead> {
        // Is an HTTP URL
        let content = if let Ok(url) = Url::parse(file.as_ref()) {
            let response = self
                .runtime
                .http
                .execute(reqwest::Request::new(reqwest::Method::GET, url))
                .await?;

            String::from_utf8(response.body.to_vec())?
        } else {
            // Is a file path

            self.runtime.file.read(file.as_ref()).await?
        };

        Ok(FileRead {
            content,
            path: file.as_ref().to_string(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    async fn get_rt() -> anyhow::Result<TargetRuntime> {
        let rt = crate::runtime::tests::init();
        let path = get_example_config();
        let content = tokio::fs::read(&path).await?;
        rt.file.write(&path, content.as_ref()).await.unwrap();
        Ok(rt)
    }

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }

    fn get_example_config() -> String {
        let mut parent = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        parent.pop();

        parent
            .join("examples/config.json")
            .to_str()
            .unwrap()
            .to_string()
    }

    #[tokio::test]
    async fn test_read_file() {
        let runtime = get_rt().await.unwrap();
        let reader = ConfigReader::init(runtime);

        let example_config = get_example_config();

        let file = reader.read_file(&example_config).await.unwrap();

        assert_eq!(file.path, example_config);
    }

    #[tokio::test]
    async fn test_read_from_url() {
        let runtime = get_rt().await.unwrap();
        let reader = ConfigReader::init(runtime);
        let expected = reader.read_file(get_example_config()).await.unwrap();

        let server = start_mock_server();

        server.mock(|when, then| {
            when.method(httpmock::Method::GET).path("/config.json");
            then.status(200).body(expected.content.as_str());
        });

        let actual = reader
            .read_file(format!("{}/config.json", server.base_url()))
            .await
            .unwrap();

        assert_eq!(expected.content, actual.content);
    }

    #[tokio::test]
    async fn test_read() {
        let runtime = get_rt().await.unwrap();
        let reader = ConfigReader::init(runtime);
        let example_config = get_example_config();

        let config = reader.read(example_config).await.unwrap();
        assert_eq!(config.server.port.unwrap(), 19194);
        assert_eq!(config.server.get_workers(), 4);
        assert_eq!(config.server.host.clone().unwrap(), "0.0.0.0");
        assert!(!config.auth.auth_db_path.is_empty());
        assert_eq!(config.auth.totp.totp_secret, "base32encodedkey");
        assert_eq!(config.auth.aes_key, "32bytebase64encodedkey");
    }
}
