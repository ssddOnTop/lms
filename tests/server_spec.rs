#[cfg(test)]
pub mod test {
    use std::borrow::Cow;
    use std::sync::Arc;
    use std::time::SystemTime;

    use anyhow::{anyhow, Result};
    use hyper::body::Bytes;
    use lms_core::http::response::Response;
    use lms_core::runtime::TargetRuntime;
    use lms_core::{EnvIO, FileIO, HttpIO, Instance};
    use reqwest::Client;
    use tokio::io::{AsyncReadExt, AsyncWriteExt};

    #[derive(Clone)]
    struct TestHttp {
        client: Client,
    }

    impl Default for TestHttp {
        fn default() -> Self {
            Self {
                client: Client::new(),
            }
        }
    }

    impl TestHttp {
        fn init() -> Arc<Self> {
            Arc::new(Self::default())
        }
    }

    #[async_trait::async_trait]
    impl HttpIO for TestHttp {
        async fn execute(&self, request: reqwest::Request) -> Result<Response<Bytes>> {
            let response = self.client.execute(request).await;
            Response::from_reqwest(
                response?
                    .error_for_status()
                    .map_err(|err| err.without_url())?,
            )
            .await
        }
    }

    #[derive(Clone)]
    struct TestFileIO {}

    impl TestFileIO {
        fn init() -> Arc<Self> {
            Arc::new(Self {})
        }
    }

    #[async_trait::async_trait]
    impl FileIO for TestFileIO {
        async fn write<'a>(&'a self, path: &'a str, content: &'a [u8]) -> anyhow::Result<()> {
            let mut file = tokio::fs::File::create(path).await?;
            file.write_all(content)
                .await
                .map_err(|e| anyhow!("{}", e))?;
            Ok(())
        }

        async fn read<'a>(&'a self, path: &'a str) -> anyhow::Result<String> {
            let mut file = tokio::fs::File::open(path).await?;
            let mut buffer = Vec::new();
            file.read_to_end(&mut buffer)
                .await
                .map_err(|e| anyhow!("{}", e))?;
            Ok(String::from_utf8(buffer)?)
        }

        async fn create_dirs<'a>(&'a self, _path: &'a str) -> Result<()> {
            Ok(())
        }
    }
    #[derive(Clone)]
    struct TestInstance {}

    impl Instance for TestInstance {
        fn now(&self) -> Result<u128> {
            Ok(SystemTime::now()
                .duration_since(SystemTime::UNIX_EPOCH)?
                .as_millis())
        }
    }

    #[derive(Clone)]
    struct TestEnv {}

    impl EnvIO for TestEnv {
        fn get(&self, _key: &str) -> Option<Cow<'_, str>> {
            Some(Cow::Borrowed("test"))
        }
    }

    pub fn init() -> TargetRuntime {
        let http = TestHttp::init();

        let file = TestFileIO::init();
        TargetRuntime {
            http,
            file,
            env: Arc::new(TestEnv {}),
            instance: Arc::new(TestInstance {}),
        }
    }
}

#[cfg(test)]
mod server_spec {
    use lms::cli::server::Server;
    use lms_auth::auth::{AuthRequest, AuthResult, SignUpDet};
    use lms_auth::local_crypto::hash_256;
    use lms_core::authdb::auth_actors::{Authority, User};
    use lms_core::config::config_module::ConfigModule;
    use lms_core::config::reader::ConfigReader;
    use reqwest::Client;

    struct TestHttp {
        url: String,
        method: reqwest::Method,
        body: String,
    }

    async fn test_req(tests: Vec<TestHttp>, config_module: ConfigModule) {
        let mut server = Server::new(config_module);
        let server_up_receiver = server.server_up_receiver();

        tokio::spawn(async move {
            server.start().await.unwrap();
        });

        server_up_receiver
            .await
            .expect("Server did not start up correctly");

        for test in tests {
            let client = Client::new();

            let task: tokio::task::JoinHandle<Result<_, anyhow::Error>> =
                tokio::spawn(async move {
                    let mut req = reqwest::Request::new(test.method, test.url.parse()?);
                    *req.body_mut() = Some(reqwest::Body::from(test.body));

                    let response = client.execute(req).await?;
                    let response_body = response.text().await?;
                    Ok(response_body)
                });

            let response = task
                .await
                .expect("Spawned task should success")
                .expect("Request should success");
            match serde_json::from_str::<AuthResult>(&response) {
                Ok(mut response) => {
                    if let Some(succ) = response.success.as_mut() {
                        succ.token = String::new(); // can't assert totp token
                    }
                    insta::assert_snapshot!(serde_json::to_string_pretty(&response).unwrap());
                }
                Err(_) => {
                    insta::assert_snapshot!(response);
                }
            }
        }
    }

    #[tokio::test]
    async fn test_server() -> anyhow::Result<()> {
        let runtime = crate::test::init();
        let reader = ConfigReader::init(runtime);
        let mut config_module = reader.read("tests/server/config.json").await?;
        if let Some(users) = config_module.extensions.users.as_mut() {
            users.insert(User {
                username: "admin".to_string(),
                name: "admin".to_string(),
                password: hash_256("admin"),
                authority: Authority::Admin,
                batch: None,
            });
        }
        let auth = config_module.extensions.auth.as_ref().unwrap();

        let auth_req = AuthRequest::new(
            "new",
            "notNewbie",
            Some(SignUpDet {
                name: "newbie".to_string(),
                authority: 2,
                admin_username: "admin".to_string(),
                admin_password: "admin".to_string(),
                batch: Some("22BCS".to_string()),
            }),
        )?;

        let auth_req_invalid_authority = AuthRequest::new(
            "new",
            "notNewbie",
            Some(SignUpDet {
                name: "newbie".to_string(),
                authority: 3,
                admin_username: "admin".to_string(),
                admin_password: "admin".to_string(),
                batch: Some("22BCS".to_string()),
            }),
        )?;

        let auth_req_invalid_pass = AuthRequest::new("new", "notNewbieIncorrectPass", None)?;

        let auth_req_no_such_user = AuthRequest::new("noSuchUser", "notNewbie", None)?;

        test_req(
            vec![
                TestHttp {
                    url: "http://localhost:19194/invalid".to_string(),
                    method: reqwest::Method::GET,
                    body: "".to_string(),
                },
                TestHttp {
                    url: "http://localhost:19194/auth".to_string(),
                    method: reqwest::Method::POST,
                    body: "invalid aes".to_string(),
                },
                TestHttp {
                    url: "http://localhost:19194/auth".to_string(),
                    method: reqwest::Method::POST,
                    body: auth.encrypt_aes("invalid body")?,
                },
                TestHttp {
                    url: "http://localhost:19194/auth".to_string(),
                    method: reqwest::Method::POST,
                    body: auth_req.into_encrypted_request(auth)?,
                },
                TestHttp {
                    url: "http://localhost:19194/auth".to_string(),
                    method: reqwest::Method::POST,
                    body: auth_req_invalid_authority.into_encrypted_request(auth)?,
                },
                TestHttp {
                    url: "http://localhost:19194/auth".to_string(),
                    method: reqwest::Method::POST,
                    body: auth_req_no_such_user.into_encrypted_request(auth)?,
                },
                TestHttp {
                    url: "http://localhost:19194/auth".to_string(),
                    method: reqwest::Method::POST,
                    body: auth_req_invalid_pass.into_encrypted_request(auth)?,
                },
            ],
            config_module,
        )
        .await;

        Ok(())
    }
}
