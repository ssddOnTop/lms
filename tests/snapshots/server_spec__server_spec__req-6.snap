---
source: tests/server_spec.rs
expression: "serde_json::to_string_pretty(&response).unwrap()"
---
{
  "error": {
    "message": "No such user found"
  },
  "code": 500
}
